File: Balanced Binary Tree/Balanced_Binary_Tree.py
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        def dfs(node):
            if not node:
                return 0

            # Compute left height
            left = dfs(node.left)
            if left == -1: return -1

            # Compute right height
            right = dfs(node.right)
            if right == -1: return -1

            # If height difference is more than 1, it's unbalanced
            if abs(left - right) > 1:
                return -1

            # Return height of current subtree
            return max(left,right) + 1

        return dfs(root) != -1



sol = Solution()
# Example 1
root1 = TreeNode(3)
root1.left = TreeNode(9)
root1.right = TreeNode(20)
root1.right.left = TreeNode(15)
root1.right.right = TreeNode(7) 
print(f"Input: root = [3,9,20,null,null,15,7]")
print(f"Output: {sol.isBalanced(root1)}\n")

# Example 2
root2 = TreeNode(1)
root2.left = TreeNode(2)
root2.right = TreeNode(2)
root2.left.left = TreeNode(3)
root2.left.right = TreeNode(3)
root2.left.left.left = TreeNode(4)
root2.left.left.right = TreeNode(4)
print(f"Input: root = [1,2,2,3,3,null,null,4,4]")
print(f"Output: {sol.isBalanced(root2)}\n")

# Example 3
root3 = TreeNode(1)
root3.left = TreeNode(2)
root3.right = TreeNode(2)
root3.left.left = TreeNode(3)
print(f"Input: root = [1,2,2,3,null,null,null]")
print(f"Output: {sol.isBalanced(root3)}\n") 

root4 = TreeNode()
print(f"Input: root = []")
print(f"Output: {sol.isBalanced(root4)}\n")

File: Java/Neetcode150/BalancedBinaryTree.java


public class BalancedBinaryTree {
      public class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;
      TreeNode() {}
      TreeNode(int val) { this.val = val; }
      TreeNode(int val, TreeNode left, TreeNode right) {
          this.val = val;
          this.left = left;
          this.right = right;
      }
  }
 

    public int height(TreeNode node){
        // An empty tree has height -1
        if (node == null) return -1;
        return 1 + Math.max(height(node.left), height(node.right));
    }
    public boolean isBalanced(TreeNode root) {
        // An empty tree satisfies the definition of a balanced tree
        if (root == null){
            return true;
        }

        // Check if subtrees have height within 1. If they do, check if the subtrees are balanced
        if (Math.abs(height(root.left) - height(root.right)) > 1){
            return false;
        }

        return (isBalanced(root.left) && isBalanced(root.right));

        
    }

    public static void main(String[] args) {
        BalancedBinaryTree solution = new BalancedBinaryTree();
        TreeNode root = solution.new TreeNode(3);
        root.left = solution.new TreeNode(9);
        root.right = solution.new TreeNode(20);
        root.right.left = solution.new TreeNode(15);
        root.right.right = solution.new TreeNode(7);

        boolean result = solution.isBalanced(root);
        System.out.println("Is the binary tree balanced? " + result);

        BalancedBinaryTree solution2 = new BalancedBinaryTree();
        TreeNode root2 = solution2.new TreeNode(1);
        root2.left = solution2.new TreeNode(2);
        root2.right = solution2.new TreeNode(2);
        root2.left.left = solution2.new TreeNode(3);
        root2.left.right = solution2.new TreeNode(3);
        root2.left.left.left = solution2.new TreeNode(4);
        root2.left.left.right = solution2.new TreeNode(4);

        boolean result2 = solution2.isBalanced(root2);
        System.out.println("Is the binary tree balanced? " + result2);

    }
}
